### PR TITLE
Refactor ReactElement in more es6 style

### PR DESCRIPTION
**Pull request by vldvel**
Add new function hasValidProp cause hasValidRef and hasValidKey are nearly the same.
And rewrite object arguments in es6 style and add destructuring for some variables.

### CODE REVIEW

File: packages/react/src/ReactElement.js [hunk 1]
```diff
@@ -21,28 +21,24 @@ const RESERVED_PROPS = {
...
-    if (hasOwnProperty.call(config, 'ref')) {
-      const getter = Object.getOwnPropertyDescriptor(config, 'ref').get;
+    if (hasOwnProperty.call(config, prop)) {
+      const getter = Object.getOwnPropertyDescriptor(config, prop).get;
       if (getter && getter.isReactWarning) {
         return false;
       }
     }
   }
-  return config.ref !== undefined;
```
Comment (gaearon) [inline packages/react/src/ReactElement.js]: Thanks, but it is important that this stays as a direct access so JS engines can better optimize what this code is doing in production.

**Comment by gaearon** [comment]
Please see my comment—this is the hottest code path, and so we intentionally sacrifice readability for keeping the code straightforward to browser engines.

File: packages/react/src/ReactElement.js (hunk 1 above)
Comment (vldvel) [reply packages/react/src/ReactElement.js]: Ohh, got it! Thanks for feedback. And what about object arguments and destructuring should I merge them?

Comment (gaearon) [reply packages/react/src/ReactElement.js]: I don't think these are important there. We generally prefer less terse code so we avoid surprises from Babel when it generates suboptimal code.

**Pull request closed without merging**